### ISSUE TITLE
proposal: databse/sql: add NullUint64 

### ISSUE THREAD

**Issue by a8m** [golang/go#47953]
Hey! I maintain an entity framework for Go named [ent](https://github.com/ent/ent), and we're having troubles handling nullable uint64 types as the official `sql` package doesn't support it. Using `NullInt64` is not enough as some databases accept values that are greater than 2^63 − 1.

Since the `sql.convertAssign` function is private my 2 options are either copy its code or use the `//go:linkname` directive.

**Comment by rsc** [comment]
Today we have:

```
type NullBool struct{ ... }
type NullByte struct{ ... }
type NullFloat64 struct{ ... }
type NullInt16 struct{ ... }
type NullInt32 struct{ ... }
type NullInt64 struct{ ... }
type NullString struct{ ... }
type NullTime struct{ ... }
```

Are you suggesting we add all the uint sizes, or just uint64? Why?

**Comment by a8m** [comment]
Hey @rsc and thanks for responding.

The `NullUint64` is required for [ent](https://github.com/ent/ent) atm to support scanning null uint64 values because using `NullInt64` returns a `value out of range` error if the value is greater than 2^63−1. 

If you see any concern with adding the rest of the unit types, it can wait for the "generics" change, because they can be implemented externally with `NullUint64`.

**Comment by kardianos** [comment]
Most SQL databases don't support uints; though some do. I would recommend we wait for a generic version to land. I was working on one when I encountered a compiler error at tip; once that is resolved I can look into it more.

I recommend against this proposal.

**Comment by a8m** [comment]
I don't think it matters if the database supports unsigned integers or not, it can also be defined as `(numeric(20,0) check (c >= 0 and c <= 2^64-1))`, and we should still be able to scan its value to the native `uint64` type without using an external package.

Currently, `Rows.Scan` (`convertAssignRows`) **does work** with `uint64`, but if the column is nullable, we have no way to work around this except for using the `sql.convertAssign` function directly or copying it.

As I mentioned above, unlike `NullByte` and `NullInt16` (that were added by me on 1.17), the `NullUint64` type can't be implemented easily outside the `sql` package, and I don't think we should be blocked for at least a year until generics will be stable.

**Comment by a8m** [comment]
Thanks for the feedback @kardianos 🙏 
 
I don't think it matters if the database supports unsigned integers or not, it can also be defined as `(numeric(20,0) check (c >= 0 and c <= 2^64-1))`, and we should still be able to scan its value to the native `uint64` type without using an external package.

Currently, `Rows.Scan` (`convertAssignRows`) **does work** with `uint64`, but if the column is nullable, we have no way to work around this except for using the `sql.convertAssign` function directly or copying it.

As I mentioned above, unlike `NullByte` and `NullInt16` (that were added by me on 1.17), the `NullUint64` type can't be implemented easily outside the `sql` package, and I don't think we should be blocked for at least a year until generics will be stable.

**Comment by rotemtam** [comment]
Thanks for opening this @a8m, on behalf of the Ent community I join your proposal 🙏

**Comment by kardianos** [comment]
@rsc I would prefer to wait for a generic Null. Unsigned types are not common in databases.
